strawberryfields/program.py

```diff
@@ -0,0 +1,897 @@
+    return
+    print_fn('i: {},  len: {}   '.format(i, len(q)), end='')
+    for x in q:
+        print_fn(x.op, ', ', end='')
+    print_fn()
+
+def _convert(func):
+    r"""Decorator for converting user defined functions to a :class:`RegRefTransform`.
+
+    This allows classical processing of measured qumodes values.
+
+    Example usage:
+
+    .. code-block:: python
+
+        @convert
+        def F(x):
+            # some classical processing of x
+            return f(x)
+
+        with eng:
+            MeasureX       | q[0]
+            Dgate(F(q[0])) | q[1]
+
+    Args:
+        func (function): function to be converted to a :class:`RegRefTransform`.
+    """
+    @functools.wraps(func)
+    def wrapper(*args):
+        "Unused docstring."
+        return RegRefTransform(args, func)
+    return wrapper
+
+
+# TODO replace this mock backend capability database
+backend_database = {
+    'fock': {
+        'New_modes': True,
+        'Delete': True,
+        'Rgate': True,
+        'Sgate': True,
+        'Dgate': True,
+        'BSgate': True,
+        'S2gate': False,  # use a decomposition
+        'MeasureFock': True,
+    },
+    'base': {
+        'New_modes': True,
+        'Delete': True,
+        'Dgate': True,
+        'Rgate': True,
+        'Sgate': True,
+        'BSgate': True,
+        'MeasureHomodyne': True,
+    }
+}
+
+
+class RegRefError(IndexError):
+    """Exception raised by :class:`Program` when it encounters an invalid register reference.
+
+    E.g. trying to apply a gate to a nonexistent or deleted subsystem.
+    """
+    pass
+
+class CircuitError(RuntimeError):
+    """Exception raised by :class:`Program` when it encounters an illegal
+    operation in the quantum circuit.
+
+    E.g. trying to use a measurement result before it is available.
+    """
+    pass
+
+class MergeFailure(RuntimeError):
+    """Exception raised by :meth:`strawberryfields.ops.Operation.merge` when an
+    attempted merge fails.
+
+    E.g. trying to merge two gates of different families.
+    """
+    pass
+
+
+class Command:
+    """Represents a quantum operation applied on specific subsystems of the register.
+
+    A Command instance is immutable once created, and can be shared between
+    several :class:`Program` instances.
+
+    Args:
+        op (~strawberryfields.ops.Operation): quantum operation to apply
+        reg (Sequence[RegRef]): Subsystems to which the operation is applied.
+            Note that the order matters here.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, op, reg, decomp=False):
+        # accept a single RegRef in addition to a Sequence
+        if not isinstance(reg, Sequence):
+            reg = [reg]
+
+        #: Operation: quantum operation to apply
+        self.op = op
+        #: Sequence[RegRef]: subsystems to which the operation is applied
+        self.reg = reg
+        #: bool: is this Command a part of a decomposition?
+        self.decomp = decomp
```

Comment:
This (was?) currently used by the engine to mark where decompositions were being applied, as a convenience to the user. This avoided a bug where calling `S2gate` would later show `S2gate`, alongside `BSgate` and `Sgate` in the `print_applied` list.

By putting the `S2gate` in a command with `decomp=False`, the engine then knows to skip it from the `print_applied` list.

Not sure if this is still needed after the refactor or not.

Reply:
Now that the compilation/optimization step is explicit I think `Command.decomp` could be removed.

Reply:
:+1: I agree.
Let's make an issue for this, and do this in a new PR as it is out of scope.